env0/resource_agent_project_assignment.go

```diff
@@ -57,138 +43,99 @@ func resourceAgentProjectAssignment() *schema.Resource {
-		return diag.Errorf("schema resource data deserialization failed: %v", err)
-	}
-
-	apaLock.Lock()
-	defer apaLock.Unlock()
-
+func resourceAgentProjectAssignmentCreateOrUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	apiClient := meta.(client.ApiClientInterface)
-	assignments, err := apiClient.ProjectsAgentsAssignments()
-	if err != nil {
-		return diag.Errorf("could not get project-agent assignments: %v", err)
-	}
 
-	for projectId, agentId := range assignments.ProjectsAgents {
-		if projectId == newAssignment.ProjectId && agentId == newAssignment.AgentId {
-			return diag.Errorf("assignment for project id %v and agent id %v already exist", projectId, agentId)
-		}
+	var assignment AgentProjectAssignment
+	if err := readResourceData(&assignment, d); err != nil {
+		return diag.Errorf("schema resource data deserialization failed: %v", err)
 	}
 
-	assignments.ProjectsAgents[newAssignment.ProjectId] = newAssignment.AgentId
+	payload := client.AssignProjectsAgentsAssignmentsPayload{
+		assignment.ProjectId: assignment.AgentId,
+	}
 
-	if _, err := apiClient.AssignAgentsToProjects(assignments.ProjectsAgents); err != nil {
-		return diag.Errorf("could not update project-agent assignments: %v", err)
+	if _, err := apiClient.AssignAgentsToProjects(payload); err != nil {
+		return diag.Errorf("failed to assign project '%s' to agent '%s': %v", assignment.ProjectId, assignment.AgentId, err)
 	}
 
-	d.SetId(newAssignment.GetId())
+	d.SetId(assignment.ProjectId)
 
 	return nil
 }
 
 func resourceAgentProjectAssignmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	apiClient := meta.(client.ApiClientInterface)
 
-	apaLock.Lock()
-	defer apaLock.Unlock()
-
 	assignments, err := apiClient.ProjectsAgentsAssignments()
 	if err != nil {
-		return diag.Errorf("could not get project-agent assignments: %v", err)
+		return diag.Errorf("failed to get project agent assignments: %v", err)
 	}
 
-	var assignment *AgentProjectAssignment
-	for projectId, agentId := range assignments.ProjectsAgents {
-		if d.Id() == GetAgentProjectAssignmentId(agentId.(string), projectId) {
-			assignment = &AgentProjectAssignment{
-				AgentId:   agentId.(string),
-				ProjectId: projectId,
-			}
-			break
-		}
+	// If there's no assignment for the project, it should default to the default agent.
+
+	assignment := AgentProjectAssignment{
+		AgentId:   assignments.DefaultAgent,
+		ProjectId: d.Id(),
 	}
 
-	if assignment == nil {
-		tflog.Warn(ctx, "Drift Detected: Terraform will remove id from state", map[string]interface{}{"id": d.Id()})
-		d.SetId("")
-		return nil
+	for projectId, agent := range assignments.ProjectsAgents {
+		if projectId == assignment.ProjectId {
+			assignment.AgentId = agent.(string)
+			break
+		}
 	}
 
-	if err := writeResourceData(assignment, d); err != nil {
+	if err := writeResourceData(&assignment, d); err != nil {
```

Comment:
Consider extracting this logic, which is shared with another action

Reply:
need to be careful with over re-using code.
Golang (unlike NodeJS) is a strongly typed language, readability is more important.

Ill have another look, but if I decided not re-use code, it was probably for a good reason.

Reply:
I could do a small reuse here... I'll push something in a couple of minutes.

Reply:
Made the required changes. Thanks for pointing it out!